Improve code comments for NMEA/SignalK plugin API

sebastien-rosset opened the pull request:
This PR adds code documentation for the plugin API, specifically for NMEA/SignalK message handling.

TwoCanPlugIn commented:
My apologies, I'm kind of confused, 

Receiving and transmiting N2K has been supported since OpenCPN 5.6, with the observable listener and GetN2000Payload and WriteCommDriverN2K API's

A number of plugins already support NMEA 2000, the built-in Dashboard plugin, Douwe's Autopilot, Engine Dashboard and Javascript plugins come to mind.

The problem with the VDR plugin is twofold, Firstly it hasn't been updated to support NMEA 2000 with the current API's, 

Secondly and more importantly there is no equivalent to PushNMEABuffer. That API allows a plugin to "send" NMEA 0183 sentences to core OpenCPN and is the basis upon which the VDR plugin is able to replay a voyage. In addition to the VDR plugin implementing the NMEA 2000 API's there would need to be a NMEA 2000 equivalent to PushNMEABuffer in order for it to replay a voyage based on NMEA 2000 data.  

An alternative approach which was implemented in the TwoCan plugin (which pre-dated OpenCPN's native supprot for NMEA 2000), is to convert NMEA 2000 messages to NMEA 0183 sentences and then use PushNMEABuffer.  That feature also allowed OpenCPN in conjunction with TwoCan to act as a bi-directional NMEA 2000/NMEA 0183 gateway. TwoCan also supported the replay of a number of different log file formats.

I will also add that the VDR plugin doesn't support SignalK. If recording and replaying voyages is important, then perhaps a more thorough design review of VDR should be undertaken?

For diagnostic purposes, there are a myriad of ways of replaying NMEA 2000 data, I think these are described in an interesting [discussion regarding logging 
](https://github.com/OpenCPN/OpenCPN/discussions/3109)

I can see a kind of benefit in that parsing of the NMEA 2000 data is performed by core OpenCPN rather than in the plugin, but perhaps that may have a negative performance impact on core OpenCPN ? Also, plugins will be limited to the PGN's that the N2K library supports. Admittedly that will be sufficient for the needs of the VDR plugin but is of little use to plugins that support proprietary PGN's such as Douwe's Autopilot. Also note that as far as I am aware the N2K libraries do not yet support a number of standard PGN's such as 129808 for DSC transmissions, 127233 MOB etc.

Also a note regarding the comment made about the registration of transmitted PGN's. This is not part of the NMEA 2000 standard. It is a requirement simply for the Actisense NGT-1. It is not used by other adapters such as Yacht Devices, or SocketCAN interfaces (presumably OpenCPN treats this as a NOP for those interfaces).

sebastien-rosset commented:
> My apologies, I'm kind of confused,
> 
> Receiving and transmiting N2K has been supported since OpenCPN 5.6, with the observable listener and GetN2000Payload and WriteCommDriverN2K API's
> 
> A number of plugins already support NMEA 2000, the built-in Dashboard plugin, Douwe's Autopilot, Engine Dashboard and Javascript plugins come to mind.
> 
> The problem with the VDR plugin is twofold, Firstly it hasn't been updated to support NMEA 2000 with the current API's,
> 
> Secondly and more importantly there is no equivalent to PushNMEABuffer. That API allows a plugin to "send" NMEA 0183 sentences to core OpenCPN and is the basis upon which the VDR plugin is able to replay a voyage. In addition to the VDR plugin implementing the NMEA 2000 API's there would need to be a NMEA 2000 equivalent to PushNMEABuffer in order for it to replay a voyage based on NMEA 2000 data.
> 
> An alternative approach which was implemented in the TwoCan plugin (which pre-dated OpenCPN's native supprot for NMEA 2000), is to convert NMEA 2000 messages to NMEA 0183 sentences and then use PushNMEABuffer. That feature also allowed OpenCPN in conjunction with TwoCan to act as a bi-directional NMEA 2000/NMEA 0183 gateway. TwoCan also supported the replay of a number of different log file formats.
> 
> I will also add that the VDR plugin doesn't support SignalK. If recording and replaying voyages is important, then perhaps a more thorough design review of VDR should be undertaken?
> 
> For diagnostic purposes, there are a myriad of ways of replaying NMEA 2000 data, I think these are described in an interesting [discussion regarding logging ](https://github.com/OpenCPN/OpenCPN/discussions/3109)
> 
> I can see a kind of benefit in that parsing of the NMEA 2000 data is performed by core OpenCPN rather than in the plugin, but perhaps that may have a negative performance impact on core OpenCPN ? Also, plugins will be limited to the PGN's that the N2K library supports. Admittedly that will be sufficient for the needs of the VDR plugin but is of little use to plugins that support proprietary PGN's such as Douwe's Autopilot. Also note that as far as I am aware the N2K libraries do not yet support a number of standard PGN's such as 129808 for DSC transmissions, 127233 MOB etc.
> 
> Also a note regarding the comment made about the registration of transmitted PGN's. This is not part of the NMEA 2000 standard. It is a requirement simply for the Actisense NGT-1. It is not used by other adapters such as Yacht Devices, or SocketCAN interfaces (presumably OpenCPN treats this as a NOP for those interfaces).

Thank you for all these details. I didn't realize the API for N2K support is quite different from SignalK and NMEA0183. I thought `WANTS_NMEA_SENTENCES`, `WANTS_NMEA_EVENTS`, and `WANTS_AIS_SENTENCES` are for NMEA0183, and that both N2k and SignalK was missing.
I was planning on updating the VDR plugin to support both N2K and SignalK. It's easier to support one in separate PRs. I have already improved it to support preferences, replay speed, pause, restart, slider, and both CSV and raw 0183 output. I will read your comments and the code in more details to understand how N2K messages are passed to plugins.

rgleason commented:
Thank you for bringing these questions up Steve, so there is some chance for us to have a versatile VDR that will play, store as file and record nmea0183, n2k and signalK at some future date. 

You raise some good points which will probably require a Plugin API change/additions to handle n2k and signalk, unless there is double translation done in the plugin itself. using nmea0183 as the common send-recieve bus.  Although TwoCan does this quite cleverly, wouldn't it be better to have the tool built-in to the API and OpenCPN?  I think this topic should be added to the OpenCPN discussions section, so a plan can be forthcoming.

Meanwhile, VDR is what we have right now, and Sebastien has done some good work cleaning it up so it can be used in many ways.  It needs to be pushed up to plugins and then worked on further.  I think adding n2k and signalK to VDR would make a great utility just as useful as nmeaconvert. 

Take small steps, try it and discuss.  
Thank you Sebastien!

sebastien-rosset commented:
> Receiving and transmiting N2K has been supported since OpenCPN 5.6, with the observable listener and GetN2000Payload and WriteCommDriverN2K API's

Looking at `GetN2000Payload`, how does a VDR plugin subscribe to all N2K messages such that it can write N2K messages to a file, even proprietary messages? Shouldn't the VDR plugin be able to serialize all N2K messages?

sebastien-rosset commented on include/ocpn_plugin.h:
```diff
@@ -1845,13 +1949,36 @@ extern DECL_EXP CommDriverResult WriteCommDriverN2K(
+ * intends to transmit on an NMEA 2000 network.
+ *
+ * This function is part of the NMEA 2000 "plug and play" capability.
+ * Before any device can transmit messages on an NMEA 2000 network, it must
```
@TwoCanPlugIn 

The previous comments were:

> Special NMEA2000 requirements.  NMEA2000 bus protocol device management requires that devices writing on the bus must inform all bus listeners of the specific PGNs that may be transmitted by this device.

Are you saying this is not true? I was trying to improve the code comments by providing the short description (purpose) and then rephrasing the existing comments.

sebastien-rosset commented:
> Thank you for bringing these questions up Steve, so there is some chance for us to have a versatile VDR that will play, store as file and record nmea0183, n2k and signalK at some future date.

My thought was to further enhance the VDR preferences and add a section with 3 checkboxes. The user could select more than one item.
☑ Record NMEA 0183 messages
☑ Record NMEA 2000 messages
☑ Record SignalK messages

There would be additional options to select the output format.

TwoCanPlugIn commented:
Several answers here.

> Special NMEA2000 requirements. NMEA2000 bus protocol device management requires that devices writing on the bus must inform all bus listeners of the specific PGNs that may be transmitted by this device.

Not exactly true, a slightly more nuanced answer is required. First of all separate "smart" devices such as the Actisense NGT-1 from "dumb" devices such as any of the SocketCAN adapters available on Linux, or any of the adapters supported by TwoCan. 

Smart devices perform the NMEA 2000 network management functions such as performing the address claim dance and responding to a few protocol requests initiated through PGN's 59904 and 126208. They also perform  the re-assembly of NMEA Fast Messages. On the other hand dumb devices don't perform this functionality, it is all performed in this case by OpenCPN (or for users of TwoCan, within the plugin). There is one PGN,  126464 in which a device can indicate what PGN's it receives and transmits, however both supporting and responding accurately for PGN 126464 is not mandatory and does not preclude the device from participating on the network. 

It is only the NGT-1 that requires software using it, to register the list of PGN's it transmits and this is only for the NGT-1 to prevent software from "misusing" the network and to ensure that the NGT-1 continues to maintain its NMEA 2000 Certification status. This PGN registration is not a requirement of the NMEA 2000 standard.

>Looking at GetN2000Payload, how does a VDR plugin subscribe to all N2K messages such that it can write N2K messages to a file, even proprietary messages.

If the role of the VDR plugin is to record and replay voyages, then there is only a limited number of PGN's (or for that matter SignalK values) that are required. There is little point for VDR to record a proprietary PGN that it cannot parse.

For diagnostic purposes, as outlined in the discussion linked to previously, there are a number of tools that can be used to log NMEA 2000 messages, on Linux for users with SocketCAN interfaces the candump utility, for users of Actisense and Yacht Devices adapters, each of those have logging facilities.

Nonetheless if you want to receive all (or as far as the plugin is concerned all PGN's that the plugin "knows"), then just register listeners for each PGN and have a single handler that can then record/parse each PGN as it sees fit.
Find below snippets of example code I've used in my own network plugin.

`std::map<unsigned int, wxString> parameterGroupNumbers = {`
`	{59392, "ISO Acknowledgement"},`
`	{59904, "ISO Request"},`
`	{60160, "ISO Transport Protocol, Data Transfer"},`
`	{60416, "ISO Transport Protocol, Connection Management"},`
`	{60928, "ISO Address Claim"},`
`	{61184, "Manufacturer Proprietary Single Frame"},`
`	{65280, "Manufacturer Proprietary Single Frame"},`
`	{65305, "Manufacturer Proprietary Single Frame (B&G AC12 Autopilot Status)"},`
`	{65309, "Manufacturer Proprietary Single Frame (B&G WS320 Battery Status)"},`
`	{65312, "Manufacturer Proprietary Single Frame (B&G WS320 Wireless Status)"},`
`	{65340, "Manufacturer Proprietary Single Frame (B&G AC12 Autopilot Mode)"},`
`	{65341, "Manufacturer Proprietary Single Frame (B&G AC12 Wind Angle)"},'`
`        ......`
`       }`

`int NetworkPlugin::Init(void) {`

`.......`
`// Listen for all NMEA 2000 Messages`
`wxDEFINE_EVENT(EVT_N2K, ObservedEvt);`
`for (const auto it : parameterGroupNumbers) {`
`	wxLogMessage("Network Plugin, Added Listener: %d, %s", it.first, it.second);`
`	listeners.push_back(std::move(GetListener(NMEA2000Id(it.first), EVT_N2K, this)));`
`}`

`Bind(EVT_N2K, [&](ObservedEvt ev) {`
`	HandleN2K(ev);`
`});`
`.......`
`}`

`void NetworkPlugin::HandleN2K(ObservedEvt ev) {`
`	std::vector<uint8_t>payload = GetN2000Payload(NMEA2000Id(123456), ev);`
`	ParseMessage(payload);`
`}`
 
Where ParseMessage extracts the PGN, Source & Destination Addresses, Priority and Payload from the raw data.

> I didn't realize the API for N2K support is quite different from SignalK and NMEA0183

It's kind of changed with OpenCPN Plugin API's 1.18 and 1.19. With 1.18 (albeit a few bugs), they are all supported with the Observable Listener model which provides an alternative mechanism to SetNMEASentence and SetPluginMessage for receiving NMEA 0183 and SignalK data respectively. With API 1.19 the Observer Listener model has been extended to support both REST and OpenCPN Plugin Messages. You might be interested in my updated and unreleased [Racing plugin](https://github.com/TwoCanPlugIn/RacingPlugin/tree/V2) which illustrates the use of both the old and new ways of retrieving NMEA 0183, SignalK and Plugin Messages in addition to NMEA 2000. In many respects, I updated it to make it a "better" demo plugin than the the built-in Demo plugin that I know you've worked on.

I think the updates you've made to VDR are great and your idea of allowing the user to select the different protocols (NMEA 0183, NMEA 2000 and SignalK) makes a lot of sense. The major impediment for VDR in supporting the replay of either NMEA 2000 or SignalK is that other than some kludgy workarounds (Eg. using virtual serial ports or additional network connections) there is no API that allows a plugin to write NMEA 2000 or SignalK to core OpenCPN. Plugins can transmit NMEA 2000 using the WriteCommDriverN2K API, however it is sent over OpenCPN NMEA 2000 connections and not consumed by core OpenCPN.   

I don't think there is any need to consider the Garmin protocol. While there's some disagreement between myself and others as to whether the Garmin protocol is a "protocol" (I believe it is), I have no idea how many users make use of it and consequently if it is worth including in your list nor how it would be plumbed in. It really only provides real time position data, which would be normally be recorded by enabling tracks.

leamas commented:
Some additional remarks:

1. I'm working with a new API which basically returns all message types received or transmitted so far.   Plan is  have it available in the plugin API as well.
2. A fully usable logging solution should probably adhere to #3667. It might be necessary to plumb this into the drivers.
3. The omission of N2K messages in the "old" interface is on purpose: The interface is slow, and N2k enables far higher data rates than 0183.  Enabling those rates on the slow, old interface obviously carries risks.

sebastien-rosset commented:
> Several answers here.

I've removed the N2K handler and added/improved code comments based on the feedback. Thank you!

> If the role of the VDR plugin is to record and replay voyages, then there is only a limited number of PGN's (or for that matter SignalK values) that are required. There is little point for VDR to record a proprietary PGN that it cannot parse.

A VDR can serve many purposes, including:
1. Post-voyage analysis of vessel and crew performance
2. Analyzing weather patterns and their effects on vessel behavior
3. Research and training purposes
4. System Troubleshooting
5. Training simulators using real-world data
6. Testing software with real data patterns
7. Building historical databases

Ideally, users should have the option to enable comprehensive recording rather than selective, as future use cases may be impossible to predict. For example, if someone wants to implement advanced analytics or machine learning, they would most likely want to have all the data.

> I think the updates you've made to VDR are great and your idea of allowing the user to select the different protocols (NMEA 0183, NMEA 2000 and SignalK) makes a lot of sense. The major impediment for VDR in supporting the replay of either NMEA 2000 or SignalK is that other than some kludgy workarounds (Eg. using virtual serial ports or additional network connections) there is no API that allows a plugin to write NMEA 2000 or SignalK to core OpenCPN. Plugins can transmit NMEA 2000 using the WriteCommDriverN2K API, however it is sent over OpenCPN NMEA 2000 connections and not consumed by core OpenCPN.

IMO, the VDR implementation for recording can be done even if OpenCPN does not support replaying N2K or SignalK messages. This makes it possible to start recording data (which can be analyzed offline), and then later when message injection capabilities are available, a follow-up VDR PR supports replay.

TwoCanPlugIn commented:
You've clearly given some thought to the different purposes that logging provides and the work that you have done in maintaining the current VDR plugin and the enhancements that you are proposing will be appreciated by many.

My only caution is that I personally believe in using the right tool for the job. 

I don't believe that the majority of end users are interested in diagnostics or decoding the various protocols. Powerboat users and non-racing sailors are unlikely to be interested in logging performance data, calibrating transducers, making comparisons to their polar tables or logging data every few seconds.  I am happy to be proven wrong!

Certainly data errors such as NMEA 0183 sentence or checksum errors should at least be logged, or made easier to identify. On the other hand NMEA 2000 is usually quite reliable, gross errors are usually caused by network faults, protocol errors are rare, generally vendors have implemented the standards correctly and the few faults that do occur usually require byte/bit level analysis.

For performance oriented logging, if you're not already familiar with it, have a look at how SignalK integrates with InfluxDB and Grafana to provide analysis tools. As the SQLite  libraries are included in O, note that Grafan has a SQLite plugin. 

For diagnostics, consider how best to use the tools readily available such as canboat for NMEA 2000, (even Tony's Javascript plugin makes use of canboatjs to parse NMEA 2000 PGN's) and how best to "feed" the tools. In general, candump is widely used for NMEA 2000.

I really think that a lot more thought needs to be given to logging, alarms, alerts but influencing the direction that OpenCPN takes is well beyond my pay grade.

sebastien-rosset commented:
Thank you for the thoughtful feedback. I understand your points about using the right tool for the job and the value of specialized solutions like SignalK with InfluxDB for advanced analysis. However, I believe there's still significant value in providing simple, file-based logging capabilities:

1. Not all users have SignalK installed or want to maintain a database system
1. Simple file logging provides an accessible entry point for users who want to:
    1. Review their route and conditions after a trip
    2. Compare basic performance data between different sails or conditions
    3. Export data to other tools they're familiar with
    
I agree that a broader discussion about OpenCPN's logging, alarms, and alerts infrastructure would be valuable. However, I don't have enough established experience and contributions to undertake such a big project. For now, I think I can work on smaller contributions that have a reasonable chance to land, as opposed to a large project that will most likely take a year or more. Given these timelines, I believe focusing on targeted, manageable improvements to existing functionality is the most productive approach for now. For reference, I currently have small bug fix PRs pending, including a fix for a segmentation fault that's less than 10 lines of code, and even for these PRs, I'm a bit uncertain whether this will move forward or not.

Would you be open to moving forward with these targeted enhancements to VDR while keeping in mind the potential for future integration with more sophisticated analysis tools?

sebastien-rosset commented:
> It's kind of changed with OpenCPN Plugin API's 1.18 and 1.19.

Is it intentional that [ocpn_plugin.h](https://github.com/OpenCPN/OpenCPN/blob/master/include/ocpn_plugin.h#L67-L68) still sets the most recent version to 1.18 in master HEAD? I see APIs for version 19 are defined, but `API_VERSION_MINOR` is still set to `18`.

```c++
#define API_VERSION_MAJOR 1
#define API_VERSION_MINOR 18
```

For VDR to use the listener, it looks like we need https://github.com/OpenCPN/OpenCPN/blob/master/include/ocpn_plugin.h#L1870-L1873

bdbcat commented:
API_VERSION_MINOR = 118 is a bug.  Thanks for the catch.
But it is only informational.  The ocpn_plugin.h in the core master contains the 119 API definitions, and current ocpn core master implements the exposed 119 functions.
Most modern ocpn plugins include the opencpn-lib submodule, rather than including the ocpn_plugin.h header in the plugin's source tree.  But VDR is a very old plugin, back when we did all this the hard way.

sebastien-rosset commented:
> Most modern ocpn plugins include the opencpn-lib submodule, rather than including the ocpn_plugin.h header in the plugin's source tree. But VDR is a very old plugin, back when we did all this the hard way.

The CMakeLists.txt shows that the plugin is actually using the modern approach:
```
add_subdirectory("opencpn-libs/${PKG_API_LIB}")
target_link_libraries(${CMAKE_PROJECT_NAME} ocpn::api)
```

The git submodule status confirms that opencpn-libs is properly set up as a submodule:
```
git submodule status
 facd6f684e8b851619a100e5601626e2002e1d42 ../opencpn-libs (heads/main)
 ```
 
 We have duplicate header files in both:
```
./libs/api-17/ocpn_plugin.h (old way)
./opencpn-libs/api-17/ocpn_plugin.h (new way)
```

If I remove the `vdr_pi/libs` directory, I'm able to build the VDR plugin. https://github.com/rgleason/vdr_pi/pull/48

bdbcat commented:
That's good.  I missed the update to use opencpn-libs somewhere along the way.
I recommend using the new way, especially since it is already in place.  And API117 is fully stable.

sebastien-rosset commented:
I've raised PR for updating the API version: https://github.com/OpenCPN/OpenCPN/pull/4268

sebastien-rosset commented:
> That's good. I missed the update to use opencpn-libs somewhere along the way.
I recommend using the new way, especially since it is already in place. And API117 is fully stable.

Unless I misunderstood, @TwoCanPlugIn  was saying API 1.19 is more stable
https://github.com/OpenCPN/OpenCPN/pull/4265#issuecomment-2543839373

> It's kind of changed with OpenCPN Plugin API's 1.18 and 1.19. With 1.18 (albeit a few bugs), they are all supported with the Observable Listener model which provides an alternative mechanism to SetNMEASentence and SetPluginMessage for receiving NMEA 0183 and SignalK data respectively.

leamas commented:
I have some remarks pn these comments , intend to review them later when I find some time.

TwoCanPlugIn commented:
> Not all users have SignalK installed or want to maintain a database system

Sorry, you missed my point, which is that an appropriate database such as InfluxDB is good for storing & retrieving time based data. You can have a separate installation of InfluxDB and accessed using the C/C++ API's without any need for SignalK. For performance oriented sailors, perhaps logging of performance related data should be a feature of Tactics or Dashboard Tactics plugins rather than the VDR plugin?
  
> Would you be open to moving forward with these targeted enhancements to VDR while keeping in mind the potential for future integration with more sophisticated analysis tools?

I'm not the maintainer of VDR so certainly feel free to maintain and add enhancements as you see fit. I'm sure Rick and the wider OpenCPN community will appreciate the improvements that you are suggesting. In the past, I have just made a few suggestions to Rick, I have not contributed any code. (I have enough problems maintaining my few plugins!)

> Unless I misunderstood, @TwoCanPlugIn was saying API 1.19 is more stable

I think stability is more a question of "time". I think Dave's comment that API 1.17 is "more stable" is simply a reflection that it has been around longer, has had more usage and consequently any bugs have been found. API 1.18 and 1.19 are much newer and have had less usage. Other than unit tests, I think the Racing Plugin is the only plugin that has used API 1.19 and the Racing Plugin is certainly is not in widespread usage.

sebastien-rosset commented:
> Sorry, you missed my point, which is that an appropriate database such as InfluxDB is good for storing & retrieving time based data. You can have a separate installation of InfluxDB and accessed using the C/C++ API's without any need for SignalK.

I get it, yes I agree it would be useful to leverage a time series database. I think both are valuable, i.e., providing some basic care to the VDR plugin with updates, and advanced analytics. Perhaps at some point I can contribute to (2), but right now I need to start with simpler contributions. On a related note, it could be valuable to label some OpenCPN issues with "Good first issue" to guide new contributors.

rgleason commented:
Point of Information
> ..separate installation of InfluxDB and accessed using the C/C++ API's without any need for SignalK. For performance oriented sailors, perhaps logging of performance related data should be a feature of Tactics or Dashboard Tactics plugins rather than the VDR plugin?

Dashboard-Tactics version in rgleason/dashboard-tactics  branch "mnt___"  is the version that we build. Petri, the author froze that branch and continued development of a more advanced version of Dashboard-tactics that uses influxDB.  At the time the features were very advanced and would only work in certain OS that supported it.  I have not tried to update that "master" branch with a frontend for compiling because I am sure that other changes will be needed and I am not the one to do this.

The point is, that InfluxDB has already been used on an advanced version of Dashboard-Tactics, and we know that it worked.

rgleason commented:
Sebastien   Should this be merged now?

sebastien-rosset commented:
> Sebastien Should this be merged now?

Yes, thank you!

rgleason commented:
I realize now that this is a PR to opencpn.  I do not have authorization for those merges. Dave or Alex.

sebastien-rosset commented:
I've been able to use the plugin API version 1.19 with `GetListener`. It works on Linux, MacOS and compiles on Windows (haven't tested), but the Android build fails. It looks like the Android library does not include the `GetListener` function.

bdbcat commented:
Correct.  Android production includes API118.

leamas commented:
@sebastien-rosset : I have made a  quick review^

sebastien-rosset commented:
> @sebastien-rosset : I have made a quick review^

Are you asking for changes?

leamas commented:
Yes, I'm asking for changes. If you don't want to do them I can apply them after merging, but the customary is that the reviewee makes the changes proposed by the reviewer.

sebastien-rosset commented:
> Yes, I'm asking for changes. If you don't want to do them I can apply them after merging, but the customary is that the reviewee makes the changes proposed by the reviewer.

Of course I can make the changes. I see you have made the following comments, but I'm unclear what you want me to change based on these remarks. I don't see other comments from you asking for specific changes.

> Some additional remarks:
I'm working with a new API which basically returns all message types received or transmitted so far. Plan is have it available in the plugin API as well. EDIT: https://github.com/OpenCPN/OpenCPN/pull/4267
A fully usable logging solution should probably adhere to https://github.com/OpenCPN/OpenCPN/issues/3667. It might be necessary to plumb this into the drivers.
The omission of N2K messages in the "old" interface is on purpose: The interface is slow, and N2k enables far higher data rates than 0183. Enabling those rates on the slow, old interface obviously carries risks.

leamas commented:
> I don't see other comments from you asking for specific changes

I have made five comments in the diff. For me, these are clearly visible just above our last posts, under "Leamas started a review". You cannot see them?

leamas commented:
If not elsewhere, you should see them here: https://github.com/OpenCPN/OpenCPN/pull/4265/files/4a20126c3ff2ccc1351657b7222528a04d613a39

Please disregard the other comments you have found, the are not part of the review.

sebastien-rosset commented:
> > I don't see other comments from you asking for specific changes
> 
> I have made five comments in the diff. For me, these are clearly visible just above our last posts, under "Leamas started a review". You cannot see them?

Did you click "Finish Review"? I think comments are not visible until you click that button.

<img width="627" alt="image" src="https://github.com/user-attachments/assets/7914765e-90c7-437a-9d1d-693017fe2943" />

leamas commented:
> Did you click "Finish your Review"?

Not exactly that button, but I had indeed forgotten to push one. Better now?

sebastien-rosset commented:
> > Did you click "Finish your Review"?
> 
> Not exactly that button, but I had indeed forgotten to push one. Better now?

Yes,  I see them now. Thanks!

sebastien-rosset commented on include/ocpn_plugin.h:
```diff
@@ -1845,13 +1934,48 @@ extern DECL_EXP CommDriverResult WriteCommDriverN2K(
+ *    - Request specific data from the transmitting device.
+ *    - Properly handle network address claims.
+ *
+ * @note
```
fixed

sebastien-rosset commented on include/ocpn_plugin.h:
```diff
@@ -429,7 +475,50 @@ class DECL_EXP opencpn_plugin {
+   * @code
+   *     // In plugin Init():
+   *     wxDEFINE_EVENT(EVT_N2K, ObservedEvt);
+   *     // Register listener for PGN 130306 (Wind Data)
```
fixed

sebastien-rosset commented on model/include/model/plugin_comm.h:
```diff
@@ -50,6 +50,18 @@ void SendPreShutdownHookToPlugins();
+ *
+ * @param sentence A complete NMEA 0183 sentence including delimiters and
+ * checksum
+ */
```
done

sebastien-rosset commented on include/ocpn_plugin.h:
```diff
@@ -80,28 +80,74 @@ class wxGLCanvas;
 //---------------------------------------------------------------------------------------------------------
+/** Receive callbacks to render custom overlay graphics on the chart.
+    Used for drawing additional navigation data, markers, or custom
+   visualizations. */
```
I've added `OnToolbarToolCallback` for `WANTS_TOOLBAR_CALLBACK`. For others, I agree it would be useful to provide more details. Perhaps in a follow-up PR?

sebastien-rosset commented on include/ocpn_plugin.h:
```diff
@@ -429,7 +475,50 @@ class DECL_EXP opencpn_plugin {
+   * @param sentence The NMEA 0183 sentence
+   *
+   * @note Starting with API 1.18, the Observable Listener model provides an
+   * alternative mechanism for receiving NMEA data. New plugins should consider
```
done

leamas commented:
Merged. Thanks for your work, on this and and elsewhere!

sebastien-rosset commented:
> I have made five comments in the diff. For me, these are clearly visible just above our last posts, under "Leamas started a review". You cannot see them?

Do you have other unpublished comments in other PRs by any chance?
